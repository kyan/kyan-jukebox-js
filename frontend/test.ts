#!/usr/bin/env bun
/**
 * Simplified Bun Test Runner with Jest Compatibility
 *
 * Uses Bun's Jest compatibility mode to run existing Jest-style tests
 * without requiring complex migration of test files.
 */

import { spawnSync } from 'child_process'
import { existsSync } from 'fs'

interface TestOptions {
  watch?: boolean
  coverage?: boolean
  verbose?: boolean
  silent?: boolean
  bail?: boolean
  watchAll?: boolean
}

class BunJestRunner {
  private options: TestOptions = {}

  constructor(args: string[] = []) {
    this.parseArgs(args)
  }

  private parseArgs(args: string[]): void {
    for (let i = 0; i < args.length; i++) {
      const arg = args[i]

      switch (arg) {
        case '--watch':
        case '-w':
          this.options.watch = true
          break
        case '--watchAll':
          this.options.watchAll = true
          break
        case '--watchAll=false':
          this.options.watchAll = false
          this.options.watch = false
          break
        case '--coverage':
          this.options.coverage = true
          break
        case '--verbose':
          this.options.verbose = true
          break
        case '--silent':
          this.options.silent = true
          break
        case '--bail':
          this.options.bail = true
          break
        case '--forceExit':
        case '--detectOpenHandles':
        case '--maxWorkers=1':
          // Legacy Jest options - ignore as Bun handles these automatically
          break
      }
    }

    // Default watch behavior
    if (
      !process.env.CI &&
      this.options.watchAll === undefined &&
      this.options.watch === undefined
    ) {
      this.options.watch = true
    }
  }

  private setupEnvironment(): void {
    // Set test environment variables
    process.env.NODE_ENV = 'test'
    process.env.BABEL_ENV = 'test'
    process.env.PUBLIC_URL = ''
  }

  private async runPreflightChecks(): Promise<void> {
    if (this.options.silent) return

    console.log('üß™ Running preflight checks...\n')

    // Type checking
    if (existsSync('./tsconfig.json')) {
      console.log('üìù Type checking...')
      try {
        const typeCheck = spawnSync('bun', ['run', 'type-check'], {
          stdio: 'inherit'
        })

        if (typeCheck.status !== 0) {
          console.error('‚ùå Type check failed')
          process.exit(1)
        } else {
          console.log('‚úÖ Type check passed')
        }
      } catch (error) {
        console.warn('‚ö†Ô∏è Type checking skipped:', error.message)
      }
    }

    // Linting
    console.log('üîç Linting...')
    try {
      const lint = spawnSync('bun', ['run', 'lint'], {
        stdio: 'inherit'
      })

      if (lint.status !== 0) {
        console.error('‚ùå Linting failed')
        process.exit(1)
      } else {
        console.log('‚úÖ Linting passed')
      }
    } catch (error) {
      console.warn('‚ö†Ô∏è Linting skipped:', error.message)
    }

    // Format check
    console.log('üíÖ Format checking...')
    try {
      const format = spawnSync('bun', ['run', 'format:check'], {
        stdio: 'inherit'
      })

      if (format.status !== 0) {
        console.error('‚ùå Format check failed')
        console.log('üí° Run: bun run format')
        process.exit(1)
      } else {
        console.log('‚úÖ Format check passed')
      }
    } catch (error) {
      console.warn('‚ö†Ô∏è Format checking skipped:', error.message)
    }

    console.log('')
  }

  private buildJestCommand(): { command: string; args: string[] } {
    const args: string[] = []

    // Test pattern - Jest will find spec files automatically
    args.push('--testMatch', '<rootDir>/src/**/*.{spec,test}.{js,jsx,ts,tsx}')

    // Watch mode
    if (this.options.watchAll) {
      args.push('--watchAll')
    } else if (this.options.watch) {
      args.push('--watch')
    } else {
      args.push('--watchAll=false')
    }

    // Coverage
    if (this.options.coverage) {
      args.push('--coverage')
    }

    // Verbose mode
    if (this.options.verbose) {
      args.push('--verbose')
    }

    // Silent mode
    if (this.options.silent) {
      args.push('--silent')
    }

    // Bail on first failure
    if (this.options.bail) {
      args.push('--bail')
    }

    // Force exit and handle detection for CI
    if (process.env.CI || !this.options.watch) {
      args.push('--forceExit', '--detectOpenHandles', '--maxWorkers=1')
    }

    return { command: 'bunx', args: ['jest', ...args] }
  }

  async run(): Promise<void> {
    if (!this.options.silent) {
      console.log('üèÉ‚Äç‚ôÇÔ∏è Bun Jest Test Runner\n')
    }

    // Setup test environment
    this.setupEnvironment()

    // Run preflight checks (validation)
    await this.runPreflightChecks()

    // Build and run Jest command through Bun
    const { command, args } = this.buildJestCommand()

    if (this.options.verbose && !this.options.silent) {
      console.log('üìã Command:', [command, ...args].join(' '))
    }

    if (!this.options.silent) {
      console.log('üß™ Running tests...\n')
    }

    try {
      const result = spawnSync(command, args, {
        stdio: 'inherit',
        cwd: process.cwd(),
        env: {
          ...process.env,
          // Ensure Bun uses Jest compatibility mode
          FORCE_COLOR: '1'
        }
      })

      if (result.status !== 0) {
        if (!this.options.silent) {
          console.error('\n‚ùå Tests failed')
        }
        process.exit(result.status || 1)
      } else {
        if (!this.options.silent) {
          console.log('\n‚úÖ All tests passed!')
        }
      }
    } catch (error) {
      console.error('‚ùå Test runner failed:', error.message)

      if (!this.options.silent) {
        console.log('\nüí° Troubleshooting:')
        console.log('   ‚Ä¢ Ensure Bun v1.0+ is installed: bun --version')
        console.log('   ‚Ä¢ Try: bun install (to ensure dependencies are correct)')
        console.log('   ‚Ä¢ Check Jest configuration in package.json')
      }

      process.exit(1)
    }
  }
}

// CLI execution
if (import.meta.main) {
  const args = process.argv.slice(2)

  if (args.includes('--help') || args.includes('-h')) {
    console.log(`
Bun Jest Test Runner

Usage:
  bun test.ts [options]

Options:
  --watch, -w          Watch files for changes
  --watchAll           Watch all files
  --watchAll=false     Disable watch mode (for CI)
  --coverage           Generate coverage report
  --verbose            Verbose output
  --silent             Suppress output except errors
  --bail               Stop on first test failure

Examples:
  bun test.ts                           # Run all tests (watch mode in dev)
  bun test.ts --watchAll=false         # CI mode - run once
  bun test.ts --coverage               # With coverage report
  bun test.ts --silent --watchAll=false # Silent CI mode
`)
    process.exit(0)
  }

  const runner = new BunJestRunner(args)
  runner.run().catch(error => {
    console.error('Test runner failed:', error)
    process.exit(1)
  })
}
