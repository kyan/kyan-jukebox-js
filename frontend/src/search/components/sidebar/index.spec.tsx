import { describe, it, expect } from 'bun:test'
import Search from './index'

describe('Search', () => {
  // Add a simple test to ensure the module can be imported
  describe('module import', () => {
    it('can import the Search component without hanging', () => {
      expect(Search).toBeDefined()
      expect(typeof Search).toBe('function')
    })
  })
})
