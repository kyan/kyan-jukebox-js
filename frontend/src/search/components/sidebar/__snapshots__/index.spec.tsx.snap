// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Search render valid props everything renders as expected 1`] = `
<div
  class="pushable"
>
  <div
    class="ui overlay right very wide visible sidebar sidebar-search"
    icon="labeled"
    inverted="true"
    vertical="true"
  >
    <div
      class="ui horizontal divider"
    >
      <h4
        class="ui inverted header"
      >
        Your Mix
      </h4>
    </div>
    <div
      class="ui tiny divided inverted list"
      role="list"
    >
      <div
        class="draggable search-list-item-draggable"
        draggable="true"
        title="You can drag this to sort."
      >
        <div
          class="search-list-item"
        >
          <img
            alt="Track name 2"
            class="ui tiny left floated image search-list-item__image"
            src="image2"
            title="Click to add - Track name 2 - Artist name 2"
          />
          <div
            class="content"
          >
            <div
              class="search-list-item__header"
            >
              Track name 2 - Artist name 2
            </div>
            <div
              class="search-list-item__content"
            >
              Album name 2
            </div>
          </div>
        </div>
        <span
          class="search-list-item__remove"
        >
          Remove
        </span>
      </div>
    </div>
    <button
      class="ui fluid button search-list-item__add_to_mix"
    >
      Add mix to playlist
    </button>
    <div
      class="ui horizontal divider"
    />
    <form
      class="ui inverted form"
    >
      <div
        class="field"
      >
        <label>
          SEARCH
        </label>
        <input
          aria-label="search-input"
          placeholder="track, artist or album"
          value=""
        />
      </div>
      <button
        class="ui fluid button"
        type="submit"
      >
        Find
      </button>
    </form>
    <div
      class="ui horizontal divider"
    >
      <h4
        class="ui inverted header"
      >
        Results
      </h4>
    </div>
    <div
      aria-label="Pagination Navigation"
      class="ui pagination pointing secondary search-list-pagination menu"
      role="navigation"
    >
      <a
        aria-current="false"
        aria-disabled="false"
        aria-label="Previous item"
        class="item"
        tabindex="0"
        type="prevItem"
        value="1"
      >
        ⟨
      </a>
      <a
        aria-current="true"
        aria-disabled="false"
        class="active item"
        tabindex="0"
        type="pageItem"
        value="1"
      >
        1
      </a>
      <a
        aria-current="false"
        aria-disabled="false"
        class="item"
        tabindex="0"
        type="pageItem"
        value="2"
      >
        2
      </a>
      <a
        aria-current="false"
        aria-disabled="false"
        aria-label="Next item"
        class="item"
        tabindex="0"
        type="nextItem"
        value="2"
      >
        ⟩
      </a>
    </div>
    <div
      class="ui tiny divided inverted relaxed list"
      role="list"
    >
      <div
        class="search-list-item disabled"
      >
        <img
          alt="Happiness"
          class="ui tiny disabled left floated image search-list-item__image"
          title="Click to add - Happiness - Ken Dodd"
        />
        <div
          class="content"
        >
          <div
            class="search-list-item__header"
          >
            Happiness - Ken Dodd
          </div>
          <div
            class="search-list-item__content"
          >
            Ken Dodd
          </div>
          <span
            class="search-list-item__add"
          >
            Add to mix
          </span>
        </div>
      </div>
      <div
        class="search-list-item"
      >
        <img
          alt="Chicken Song"
          class="ui tiny left floated image search-list-item__image"
          src="https://i.scdn.co/image/e0fc5e87ca6e137d4d7ef434f73edbfb4543286b"
          title="Click to add - Chicken Song - 3 Shades"
        />
        <div
          class="content"
        >
          <div
            class="search-list-item__header"
          >
            Chicken Song - 3 Shades
          </div>
          <div
            class="search-list-item__content"
          >
            Chicken Song
          </div>
          <div
            class="ui grey mini basic label track-label vote-track-label"
          >
            <i
              aria-hidden="true"
              class="thumbs down icon"
            />
            -1
          </div>
          <span
            class="search-list-item__add"
          >
            Add to mix
          </span>
        </div>
      </div>
      <div
        class="search-list-item"
      >
        <img
          alt="I Know Him So Well"
          class="ui tiny left floated image search-list-item__image"
          src="https://i.scdn.co/image/ab67616d0000b273f5d46038d259167ee9791bee"
          title="Click to add - I Know Him So Well - Geraldine McQueen"
        />
        <div
          class="content"
        >
          <div
            class="search-list-item__header"
          >
            I Know Him So Well - Geraldine McQueen
          </div>
          <div
            class="search-list-item__content"
          >
            I Know Him So Well
          </div>
          <span
            class="search-list-item__add"
          >
            Add to mix
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    class="pusher dimmed"
  />
</div>
`;
